qiskit_nature/operators/second_quantization/sparse_label_op.py

```diff
@@ -65,7 +64,7 @@ def _multiply(self, other: Number) -> SparseLabelOp:
         Raises:
             TypeError: if ``other`` is not compatible type (int, float or complex)
         """
-        if not isinstance(other, Number):
+        if not isinstance(other, (int, float, complex)):
```

Comment:
I would be curious, if you just leave this Number check here if things are ok (not quite sure what the "fun" with mypy was). If you look in PolynomialTensor there it uses Number for the instance check. For input & checking the more flexibility allows the caller flexibility. I recall when we  test things too 'precisely' we have had cases like numpy ints that fail these checks when its reasonable to have expected things to have worked.

Reply:
I'll try adding it back and see what happens!

The 'fun' with mypy shows up here if you want more details: https://github.com/Qiskit/qiskit-nature/runs/7862883533?check_suite_focus=true

Reply:
I think the aspect with iter is that its getting information about the type from data as passed to the constructor Dict[str, complex] so self_.data.items() becomes an iterable over str, complex because that was what was passed in. I would imagine changing the Dict type to [str, Number] would resolve that. As to multiply, that overrides mixin defined method right. Which is why I think in the PoynomialTensor the type was left as complex though the instance check was done on Number. As in the body then is similar too to quantum info operator implementation https://github.com/Qiskit/qiskit-terra/blob/f2d228ecd288c928d074b1f61702b96eb5b66e87/qiskit/quantum_info/operators/operator.py#L401-L414